off_moo_bench/problem/lambo/data/experiments/test/hydra_config.txt
logger:
  _target_: upcycle.logging.DataFrameLogger
  log_dir: data/experiments/test/agile_voile_0/2025-05-26_13-45-50
task:
  _target_: lambo.tasks.proxy_rfp.proxy_rfp.ProxyRFPTask
  obj_dim: 2
  log_prefix: proxy_rfp
  batch_size: 16
  max_len: 244
  max_num_edits: null
  max_ngram_size: 1
  allow_len_change: false
  num_start_examples: 512
acquisition:
  _target_: lambo.acquisitions.ehvi.NoisyEHVI
  num_samples: 2
  batch_size: 16
encoder:
  _target_: lambo.models.lm_elements.LanguageModel
  name: mlm_cnn
  model:
    _target_: lambo.models.shared_elements.mCNN
    tokenizer:
      _target_: lambo.utils.ResidueTokenizer
    max_len: 244
    embed_dim: 64
    latent_dim: 16
    out_dim: 16
    kernel_size: 5
    p: 0.0
    layernorm: true
    max_len_delta: 0
  batch_size: 32
  num_epochs: 128
  patience: 32
  lr: 0.001
  max_shift: 0
  mask_ratio: 0.125
optimizer:
  _target_: lambo.optimizers.pymoo.ModelFreeGeneticOptimizer
  _recursive_: false
  num_rounds: 32
  num_gens: 2
  seed: 0
  concentrate_pool: 1
  residue_sampler: uniform
  resampling_weight: 1.0
  algorithm:
    _target_: pymoo.algorithms.moo.nsga2.NSGA2
    pop_size: 16
    n_offsprings: null
    sampling:
      _target_: lambo.optimizers.sampler.CandidateSampler
    crossover:
      _target_: pymoo.factory.get_crossover
      name: int_sbx
      prob: 0.0
      eta: 16
    mutation:
      _target_: lambo.optimizers.mutation.LocalMutation
      prob: 1.0
      eta: 16
      safe_mut: false
    eliminate_duplicates: true
tokenizer:
  _target_: lambo.utils.ResidueTokenizer
surrogate:
  _target_: lambo.models.gp_models.MultiTaskExactGP
  max_shift: 0
  mask_size: 0
  bootstrap_ratio: null
  min_num_train: 128
  task_noise_init: 0.25
  gp_lr: 0.005
  enc_lr: 0.005
  bs: 32
  eval_bs: 16
  num_epochs: 256
  holdout_ratio: 0.2
  early_stopping: true
  patience: 32
  eval_period: 2
  out_dim: 2
  feature_dim: 16
  encoder_wd: 0.0001
  rank: null
  task_covar_prior:
    _target_: gpytorch.priors.LKJCovariancePrior
    'n': 2
    eta: 2.0
    sd_prior:
      _target_: gpytorch.priors.SmoothedBoxPrior
      a: 0.0001
      b: 1.0
  data_covar_module:
    _target_: gpytorch.kernels.MaternKernel
    ard_num_dims: 16
    lengthscale_prior:
      _target_: gpytorch.priors.NormalPrior
      loc: 0.7
      scale: 0.01
  likelihood:
    _target_: gpytorch.likelihoods.MultitaskGaussianLikelihood
    num_tasks: 2
    has_global_noise: false
    noise_constraint:
      _target_: gpytorch.constraints.GreaterThan
      lower_bound: 0.0001
seed: 0
trial_id: 0
project_name: lambo
version: v0.2.1
data_dir: data/experiments
exp_name: test
job_name: agile_voile_0
timestamp: 2025-05-26_13-45-50
log_dir: data/experiments/test
wandb_mode: online
wandb_host: https://api.wandb.ai


off_moo_bench/problem/morl/externals/baselines/baselines/her/experiment/config.py
import os

import gym
import numpy as np
from baselines import logger
from baselines.bench.monitor import Monitor
from baselines.her.ddpg import DDPG
from baselines.her.her_sampler import make_sample_her_transitions

DEFAULT_ENV_PARAMS = {
    "FetchReach-v1": {
        "n_cycles": 10,
    },
}


DEFAULT_PARAMS = {
    # env
    "max_u": 1.0,  # max absolute value of actions on different coordinates
    # ddpg
    "layers": 3,  # number of layers in the critic/actor networks
    "hidden": 256,  # number of neurons in each hidden layers
    "network_class": "baselines.her.actor_critic:ActorCritic",
    "Q_lr": 0.001,  # critic learning rate
    "pi_lr": 0.001,  # actor learning rate
    "buffer_size": int(1e6),  # for experience replay
    "polyak": 0.95,  # polyak averaging coefficient
    "action_l2": 1.0,  # quadratic penalty on actions (before rescaling by max_u)
    "clip_obs": 200.0,
    "scope": "ddpg",  # can be tweaked for testing
    "relative_goals": False,
    # training
    "n_cycles": 50,  # per epoch
    "rollout_batch_size": 2,  # per mpi thread
    "n_batches": 40,  # training batches per cycle
    "batch_size": 256,  # per mpi thread, measured in transitions and reduced to even multiple of chunk_length.
    "n_test_rollouts": 10,  # number of test rollouts per epoch, each consists of rollout_batch_size rollouts
    "test_with_polyak": False,  # run test episodes with the target network
    # exploration
    "random_eps": 0.3,  # percentage of time a random action is taken
    "noise_eps": 0.2,  # std of gaussian noise added to not-completely-random actions as a percentage of max_u
    # HER
    "replay_strategy": "future",  # supported modes: future, none
    "replay_k": 4,  # number of additional goals used for replay, only used if off_policy_data=future
    # normalization
    "norm_eps": 0.01,  # epsilon used for observation normalization
    "norm_clip": 5,  # normalized observations are cropped to this values
    "bc_loss": 0,  # whether or not to use the behavior cloning loss as an auxilliary loss
    "q_filter": 0,  # whether or not a Q value filter should be used on the Actor outputs
    "num_demo": 100,  # number of expert demo episodes
    "demo_batch_size": 128,  # number of samples to be used from the demonstrations buffer, per mpi thread 128/1024 or 32/256
    "prm_loss_weight": 0.001,  # Weight corresponding to the primary loss
    "aux_loss_weight": 0.0078,  # Weight corresponding to the auxilliary loss also called the cloning loss
}


CACHED_ENVS = {}


def cached_make_env(make_env):
    """
    Only creates a new environment from the provided function if one has not yet already been
    created. This is useful here because we need to infer certain properties of the env, e.g.
    its observation and action spaces, without any intend of actually using it.
    """
    if make_env not in CACHED_ENVS:
        env = make_env()
        CACHED_ENVS[make_env] = env
    return CACHED_ENVS[make_env]


def prepare_params(kwargs):
    # DDPG params
    ddpg_params = dict()
    env_name = kwargs["env_name"]

    def make_env(subrank=None):
        env = gym.make(env_name)
        if subrank is not None and logger.get_dir() is not None:
            try:
                from mpi4py import MPI

                mpi_rank = MPI.COMM_WORLD.Get_rank()
            except ImportError:
                MPI = None
                mpi_rank = 0
                logger.warn(
                    "Running with a single MPI process. This should work, but the results may differ from the ones publshed in Plappert et al."
                )

            max_episode_steps = env._max_episode_steps
            env = Monitor(
                env,
                os.path.join(logger.get_dir(), str(mpi_rank) + "." + str(subrank)),
                allow_early_resets=True,
            )
            # hack to re-expose _max_episode_steps (ideally should replace reliance on it downstream)
            env = gym.wrappers.TimeLimit(env, max_episode_steps=max_episode_steps)
        return env

    kwargs["make_env"] = make_env
    tmp_env = cached_make_env(kwargs["make_env"])
    assert hasattr(tmp_env, "_max_episode_steps")
    kwargs["T"] = tmp_env._max_episode_steps

    kwargs["max_u"] = (
        np.array(kwargs["max_u"])
        if isinstance(kwargs["max_u"], list)
        else kwargs["max_u"]
    )
    kwargs["gamma"] = 1.0 - 1.0 / kwargs["T"]
    if "lr" in kwargs:
        kwargs["pi_lr"] = kwargs["lr"]
        kwargs["Q_lr"] = kwargs["lr"]
        del kwargs["lr"]
    for name in [
        "buffer_size",
        "hidden",
        "layers",
        "network_class",
        "polyak",
        "batch_size",
        "Q_lr",
        "pi_lr",
        "norm_eps",
        "norm_clip",
        "max_u",
        "action_l2",
        "clip_obs",
        "scope",
        "relative_goals",
    ]:
        ddpg_params[name] = kwargs[name]
        kwargs["_" + name] = kwargs[name]
        del kwargs[name]
    kwargs["ddpg_params"] = ddpg_params

    return kwargs


def log_params(params, logger=logger):
    for key in sorted(params.keys()):
        logger.info("{}: {}".format(key, params[key]))


def configure_her(params):
    env = cached_make_env(params["make_env"])
    env.reset()

    def reward_fun(ag_2, g, info):  # vectorized
        return env.compute_reward(achieved_goal=ag_2, desired_goal=g, info=info)

    # Prepare configuration for HER.
    her_params = {
        "reward_fun": reward_fun,
    }
    for name in ["replay_strategy", "replay_k"]:
        her_params[name] = params[name]
        params["_" + name] = her_params[name]
        del params[name]
    sample_her_transitions = make_sample_her_transitions(**her_params)

    return sample_her_transitions


def simple_goal_subtract(a, b):
    assert a.shape == b.shape
    return a - b


def configure_ddpg(dims, params, reuse=False, use_mpi=True, clip_return=True):
    sample_her_transitions = configure_her(params)
    # Extract relevant parameters.
    gamma = params["gamma"]
    rollout_batch_size = params["rollout_batch_size"]
    ddpg_params = params["ddpg_params"]

    input_dims = dims.copy()

    # DDPG agent
    env = cached_make_env(params["make_env"])
    env.reset()
    ddpg_params.update(
        {
            "input_dims": input_dims,  # agent takes an input observations
            "T": params["T"],
            "clip_pos_returns": True,  # clip positive returns
            "clip_return": (1.0 / (1.0 - gamma))
            if clip_return
            else np.inf,  # max abs of return
            "rollout_batch_size": rollout_batch_size,
            "subtract_goals": simple_goal_subtract,
            "sample_transitions": sample_her_transitions,
            "gamma": gamma,
            "bc_loss": params["bc_loss"],
            "q_filter": params["q_filter"],
            "num_demo": params["num_demo"],
            "demo_batch_size": params["demo_batch_size"],
            "prm_loss_weight": params["prm_loss_weight"],
            "aux_loss_weight": params["aux_loss_weight"],
        }
    )
    ddpg_params["info"] = {
        "env_name": params["env_name"],
    }
    policy = DDPG(reuse=reuse, **ddpg_params, use_mpi=use_mpi)
    return policy


def configure_dims(params):
    env = cached_make_env(params["make_env"])
    env.reset()
    obs, _, _, info = env.step(env.action_space.sample())

    dims = {
        "o": obs["observation"].shape[0],
        "u": env.action_space.shape[0],
        "g": obs["desired_goal"].shape[0],
    }
    for key, value in info.items():
        value = np.array(value)
        if value.ndim == 0:
            value = value.reshape(1)
        dims["info_{}".format(key)] = value.shape[0]
    return dims
